src/merkle/partial_mt/mod.rs

```diff
@@ -0,0 +1,270 @@
+            nodes: BTreeMap::new(),
+            leaves: BTreeSet::new(),
+        }
+    }
+
+    /// Returns a new [PartialMerkleTree] instantiated with leaves set as specified by the provided
+    /// entries.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - If the depth is 0 or is greater than 64.
+    /// - The number of entries exceeds the maximum tree capacity, that is 2^{depth}.
+    /// - The provided entries contain multiple values for the same key.
+    pub fn with_leaves<R, I>(entries: R) -> Result<Self, MerkleError>
+    where
+        R: IntoIterator<IntoIter = I>,
+        I: Iterator<Item = (NodeIndex, RpoDigest)> + ExactSizeIterator,
+    {
+        // create an empty tree
+        let mut tree = PartialMerkleTree::new();
+
+        // check if the number of leaves can be accommodated by the tree's depth; we use a min
+        // depth of 63 because we consider passing in a vector of size 2^64 infeasible.
+        let entries = entries.into_iter();
+        let max = (1_u64 << 63) as usize;
+        if entries.len() > max {
+            return Err(MerkleError::InvalidNumEntries(max, entries.len()));
+        }
+
+        for (node_index, rpo_digest) in entries {
+            let old_value = tree.update_leaf(node_index, rpo_digest)?;
+            if old_value != Self::EMPTY_DIGEST {
+                return Err(MerkleError::DuplicateValuesForIndex(node_index.value()));
+            }
+        }
+        Ok(tree)
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns the root of this Merkle tree.
+    pub fn root(&self) -> Word {
+        self.root.into()
+    }
+
+    /// Returns the depth of this Merkle tree.
+    // TODO: maybe it's better to rename it to the `max_depth`
+    pub fn depth(&self) -> u8 {
+        self.max_depth
+    }
+
+    /// Returns a node at the specified NodeIndex.
+    ///
+    /// # Errors
+    /// Returns an error if the specified NodeIndex is not contained in the nodes map.
+    pub fn get_node(&self, index: NodeIndex) -> Result<Word, MerkleError> {
+        self.nodes
+            .get(&index)
+            .ok_or(MerkleError::NodeNotInSet(index))
+            .map(|hash| **hash)
+    }
+
+    /// Returns a value of the leaf at the specified NodeIndex.
+    ///
+    /// # Errors
+    /// Returns an error if the NodeIndex is not contained in the leaves set.
+    pub fn get_leaf(&self, index: NodeIndex) -> Result<Word, MerkleError> {
+        if !self.leaves.contains(&index) {
+            // This error not really suitable in this situation, should I create a new error?
+            Err(MerkleError::InvalidIndex {
+                depth: index.depth(),
+                value: index.value(),
+            })
+        } else {
+            self.nodes
+                .get(&index)
+                .ok_or(MerkleError::NodeNotInSet(index))
+                .map(|hash| **hash)
+        }
+    }
```

Comment:
I'm actually not sure if we need this method. But we could add something like:
```Rust
pub fn is_leaf(&self, index: NodeIndex) -> bool
```

Also, probably something like [get_leaf_depth()](https://github.com/0xPolygonMiden/crypto/blob/main/src/merkle/store/mod.rs#L203) would be useful. The signature could be:
```Rust
pub fn get_leaf_depth(&self, index: u64) -> Result<u8, MerkleError> {
    ...
}
```

Reply:
I'm not sure I understand how `get_leaf_depth` works -- what index is provided into it? Is it a "global" index of the node in the tree? Not sure how this function can be used in case of `PartialMerkleTree`.

Reply:
`get_leaf_depth` would traverse the tree from the root down according to the provided index and would stop at the first leaf it finds.

Let's say this is our Partial Merkle tree:
```
                _______ (0, 0) _______
               /                      \
         _ (1, 0) _                    D
        /          \             
       A         (2, 1)    
                  /  \  
                 B    C
```
Calling `get_leaf_depth(0)` would return the depth of `A` (i.e., 2). But also calling `get_leaf_dpeth(1)` would return the depth of `A`.

Calling `get_leaf_depth(2)` would return the depth of `B` (i.e., 3)`, and calling `get_leaf_depth(5)` would return the depth of `D` (i.e., 1).